[HUDI-7856] Adding RFC-78 for 1.0 migration from 0.x (#11514)
apache/hudi: rfc/rfc-78/rfc-78.md

re-imagination of the transactional database layer in Hudi to power continued innovation across the community in the coming 
years. It introduces lot of differentiating features for Apache Hudi. We released beta releases which was meant for 
enthusiastic developers/users to give a try of advanced features. But as we are working towards 1.0 GA, we are proposing 
a bridge release (0.16.0) for smoother migration for existing hudi users. 

## Objectives 
Goal is to have a smooth migration experience for the users from 0.x to 1.0. We plan to have a 0.16.0 bridge release asking everyone to first migrate to 0.16.0 before they can upgrade to 1.x.

- 1.x reader should be able to read 0.16.x tables w/o any loss in functionality and no data inconsistencies.
- 0.16.x should have read capability for 1.x tables w/ some limitations. For features ported over from 0.x, no loss in functionality should be guaranteed. But for new features that was introduced in 1.x, we may not be able to support all of them. Will be calling out which new features may not work with 0.16.x reader. In this case, we explicitly request users to not turn on these features till readers are completely in 1.x.

> - Document upgrade steps from 0.16.x to 1.x with limited user perceived latency. This will be auto upgrade, but document clearly what needs to be done.

Reviewer: ```suggestion
- Document steps for rolling upgrade from 0.16.x to 1.x , with minimal downtime
```

> - Downgrade from 1.x to 0.16.x documented with call outs on any functionality.

Reviewer: what would these call outs be about?

### Considerations when choosing Migration strategy
- While migration is happening, we want to allow readers to continue reading data. This means, we cannot employ a stop-the-world strategy when we are migrating. 
All the actions that we are performing as part of table upgrade should not have any side-effects of breaking snapshot isolation for readers.
- Also, users should have migrated to 0.16.x before upgrading to 1.x. We do not want to add read support for very old versions of hudi in 1.x(for eg 0.7.0). 

> - So, in an effort to bring everyone to latest hudi versions, 1.x reader will have full read capabilities for 0.16.x, but for older hudi versions, 1.x reader may not have full reader support. 

Reviewer: may be good to document what works. but best to get everyone to 0.16 or those users can choose to take a downtime and do it directly?

> - So, in an effort to bring everyone to latest hudi versions, 1.x reader will have full read capabilities for 0.16.x, but for older hudi versions, 1.x reader may not have full reader support. 

Author: updated the details

The reocmmended guideline is to upgrade all readers and writers to 0.16.x. and then slowly start upgrading to 1.x(readers followed by writers). 

Before we dive in further, lets understand the format changes:

## Format changes
### Table properties

> - Payload class ➝ payload type.

Reviewer 2: We have introduced the record merge mode `hoodie.record.merge.mode` as the single table config to control the record merging behavior (see #9894).  There is a preliminary PR on inferring record merge mode during table upgrade #11439.

Author: if its custom, from where custom configs are picked up. its a table prop in 0.x, but in 1.x? 
lets validate the flows

Reviewer 3: Might not be relared, but should `hoodie.record.merge.mode` should be a table config instead of a write config?

- New metadata partitions could be added (optionally enabled)

### MDT changes
- New MDT partitions are available in 1.x. MDT schema upgraded.
- RLI schema is upgraded to hold row position

### Timeline:
- [storage changes] Completed write commits have completed times in the file name.
- [storage changes] Completed and inflight write commits are in avro format which were json in 0.x.
- We are switching the action type for clustering from “replace commit” to “cluster”.
- Similarly, for completed compaction, we are switching from “commit” to “compaction” in an effort to standardize actions for a given write operation.

> - [storage changes] Timeline ➝ LST timeline. There is no archived timeline in 1.x

Reviewer 2: More details can be provided on two parts of the timeline where older instants are stored in the LSM tree and recent instants are stored as individual files.

Author: are you talking about 1.x reader or 0.16.x reader.

> - [In-memory changes] HoodieInstant changes due to presence of completion time for completed HoodieInstants.

Reviewer 2: @danny0405 For an Hudi table as the incremental source, have we changed the semantics of the checkpoint (e.g., `deltastreamer.checkpoint.key`) stored in the commit metadata, from the start/instant time to the completion time?  If so, how to interpret the checkpoint changes across different table versions.

Author: good point. we might need to solve this elegantly

Reviewer 3: We do not introduce the completion time based inc queries for Spark yet, but for the GA release, we might need to have a compatible solution for migrattion.

Reviewer: lets keep this issue open until we resolve one way or other. Good call out.. 

Also @nsivabalan lets refrain from mentioning class names in the RFC? and refer to changes at a conceptual level - since we are discussing storage level changes?

### Filegroup/FileSlice changes:
- Log files contain delta commit time instead of base instant time.

> - Log appends are disabled in 1.x. In other words, each log block is already appended to a new log file.

Reviewer 2: Another change to call out is rollback deletes the inflight log files and there will be no rollback block appends.

Author: sure.

Reviewer 3: +1

- File Slice determination logic for log files changed (in 0.x, we have base instant time in log files and its straight forward. In 1.x, we find completion time for a log file and find the base instant time (parsed from base files) which has the highest value lesser than the completion time of the log file).

> - Log file ordering within a file slice. (in 0.x, we use base instant time ➝l log file versions ➝ write token) to order diff log files. in 1.x, we will be using completion time to order).

Author: or anyways, we are adding filterUncommittedLogs capability in FSV to read 1.x reader. So, we should be good. but lets add tests for these.

### Log format changes:
- We have added new header types in 1.x. (IS_PARTIAL)

## Changes to be ported over 0.16.x to support reading 1.x tables
### What will be supported
- For features introduced in 0.x, and tables written in 1.x, 0.16.0 reader should be able to provide consistent reads w/o any breakage.

> ### What will not be supported

Reviewer: can we use a table to clearly capture what will be supported across different reader/writer versions

Author: sure.

Author: Added a table to call our reader writer compatibility.

- A 0.16 writer cannot write to a table that has been upgraded-to/created using 1.x without downgrading to 0.16
- For new features introduced in 1.x, we might call out that 0.16.x reader may not support reads.
- if deletion vector is enabled, 0.16.0 reader should fallback to key based merges, giving up on the performance optimization. 

> - if partial updates are enabled, 0.16.0 reader should fallback to key based merges, giving up on the performance optimization.

Reviewer 2: Partial updates need new merging logic so it may not be supported without porting the (key-based) partial merging logic to 0.16.

Author: sg. the fallback support needs to be ported to 0.16. will update the RFC. is it the same for deletion vector as well ?

- 0.16.x reader may not support functional indexes built in 1.x. Read may not break, just that the optimization may not kick in.
- 0.16.x reader may not support secondary indexes built in 1.x. Read may not break, just that the optimization may not kick in.
- 0.16.x reader may not support Completion time based log file ordering while reading log files/blocks. 
  So, if you are looking to enable NBCC for your 1.x tables, would recommend to upgrade the readers to 1.x before tuning the feature on.  

#### Completion time based read in FileGroup/FileSlice initialization and log block reads
- If we were to make MDT NBCC default in 1.0, we have to support completion time based apis and implementations in 0.16. ??

> - If not, we still have a chance to not support completion time based reads with 0.16.0.

Author: we are still flushing out few details. we are also inclining towards having one way of doing things for the simplicity it brings.

    - We might have to revisit this if we plan to make NBCC default with MDT in 1.x
- If not for completion time based read support, what will break, check [here](#Completion time based read in FileGroup/FileSlice initialization and log block reads) 

### Table properties
- payload type inference from payload class need to be ported.

### MDT changes:
- MDT schema changes need to be ported.

### Log format changes:
- New log header type need to be ported.
- For unsupported features, meaningful errors should be thrown. For eg, if DV or partial update has been enabled and if those log files are read using 0.16.0 reader, we should fail and throw a meaningful error.

### Incremental reads and time travel query: 
- Incremental reads in 1.x is expected to have some changes and design is not fully out. So, until then we have to wait to design 0.16.x read support for 1.x tables for incremental queries. 

> - Time travel query: I don't see any changes in 1.x wrt time travel query. So, as per master, we are still using commit time(i.e not completed time) to serve time travel query. Until we make any changes to that, we do not need to add any additional support to 0.16.x reader. But if we plan to make changes in 1.x, we might have to revisit this. 

Author: not yet. from what I see, we have not made any changes to 1.x reader for time travel. So, unless we make changes to that logic, I don't expect we need to port anything back to 0.16x. (the file slice determination, log file ordering logic anyways will be ported. just that no new changes specific to time travel)

## 0.16.0 ➝ 1.0 upgrade steps
This will be an automatic upgrade for users when they start using 1.x hudi library. Listing the changes we might need to do with the upgrade.
- Rewrite archival timeline to LSM based timeline.
- Do not touch active timeline, since there could be concurrent readers reading the table. So, 1.x reader should be capable of handling timeline w/ a mix of 0.x commit files and 1.x commit files.
- No changes to log reader.
- Check custom payload class in table properties and switch to payload type.

> - Trigger compaction for latest file slices. We do not want a single file slice having a mix of log files from 0.x and log files from 1.x. So, we will trigger a full compaction 

Author: yes, we are thinking we should avoid full compaction in upgrade code path. downgrade should be fine (since anyways table might be in corrupt state or something). but upgrade lies in fast path and we could be having streaming pipelines. and w/ MDT(since its MOR), we can't afford to do full compaction. Lets review the RFC to gauge the complexity we are running into and take a call.

    So, using the upgrade commit time, we should be able to use different log file comparator to order log files within a given file slice. 

## 1.0 ➝ 0.16.0 downgrade steps
Users will have to use hudi-cli to do the downgrade. Here are the steps we might need to do during downgrade.
- Similar to our upgrade, lets trigger full compaction to latest file slice so that all latest file slices only has base files. We do not want to have any file slices w/ a mix of 1.x log files and 0.x log files.
- Rewrite LSM based timeline to archived timeline. we have to deduce writer properties and introduce boundaries based on that (i.e. until which commit to archive).
- We have two options wrt handling active timeline instants:
   A: No changes to active instants. In order to support 1.x tables w/ 0.16.0 reader, already these changes might have been ported to 0.16.0 reader. But we might have to ensure downgrade from 0.16.0 to any older version of hudi does take care of rewriting the active timeline instants.

>    B: While downgrading from 1.x to 0.16.0, let’s rewrite the active timeline instants too. So that after downgrade, we can't differentiate whether table was natively created in 0.16.0 or was it upgraded to 1.x and then later downgraded. The con in Option A is not an issue here since we take care of active timeline instants during 1.x to 0.16.0 downgrade only. So, downgrade from 0.16.0 to any lower hudi version does not need to do any special handling if table was native to 0.16.0 or was it downgraded from 1.x to 0.16.0

Reviewer: again, my advice is to keep this simple and retryable - vs trying to handle a lot of corner cases inside, for ease-of-use in migration tooling.

   Our proposal is to go with Option B to ensure 0.16.0 table will be intact and not have any residues from 1.x writes. While the rewrite of active instants are happening, we could see some read failures. But 
   we do not want to have a 0.16.0 (post upgrade) table to deal w/ 1.x timeline intricacies. 
- If there are new MDT partitions (functional index, sec index), nuke them. Update table properties.
- Check latest snapshot files for log files w/ new features enabled. For eg, deletion vector, partial update. If any log files are found in latest snapshot, we have to trigger compaction for those file groups. This calls for a custom compaction strategy as well which compacts only for certain file groups.
   - In order to reduce downtime to downgrade, we are inclining to do this only for snapshot reads. If there are older file slices and users are issuing time travel or incremental reads, and if they have any of the new features enabled (deletion vector, partial update), readers could break after downgrade.

> ## Implementation details

Reviewer 2: Could we have a separate sections to discuss how all different query types, i.e., snapshot, read-optimized, incremental, CDC, time-travel, operate correctly before and after migration in different scenarios of reader and writer (0.15 -> 0.16 -> 1.0), e.g., the below scenario of independently upgrading reader and writer?
```
writer:  <--- 0.15 ------>  |  <-------- 0.16 ------> | <-------- 1.0 ----
reader: <--- 0.15 -------------->   | <-------- 0.16 -----------> | <-------- 1.0 ----
```

Author: can you clarify me something as to what you want me to cover here. 
we are only adding 1.x read support to 0.16.0. but 1.x should be able to read any tables from 0.14.0 to 1.x. So, we can't really use 0.15.0 reader to read 1.x table. So, users have to upgrade to 0.16.x first before they can read any of 1.x tables.

Reviewer 3: probably just adding a matrix table for 0.16.x and 1.x, with different write and read scenarios.

### What will be supported
- For features introduced in 0.x, and tables written in 1.x, 0.16.0 reader should be able to provide consistent reads w/o any breakage.
- 
### What will not be supported
- A 0.16 writer cannot write to a table that has been upgraded-to/created using 1.x without downgrading to 0.16. Might be obvious, but calling it out nevertheless.
- For new features introduced in 1.x, we might call out that 0.16.x reader may not support reads.

| 1.x Features written by 1.x writer        | 1.x reader   | 0.16.x Reader | 
|----------------|-------------------------------------------|-------|
| Deletion vector | supported | Falls back to key based merges giving up on perf optimization
| Partial merges/updates  | supported | Fails with clear error message stating that partial merging |
| Functional indexes | supported | Not supported. Perf optimization may not kick in|
| Secondary indexes | supported | Not supported. Perf optimization may not kick in|
| NBCC OR Completion time based log file ordering in a file slice | supported | Not supported. Will be using log file version and write token based ordering.|  

### Timeline

> - Timeline read of 1.x need to be supported.

Author: Our idea here is to not port LSM archived timeline reader to 0.16.x.
We need to fix 1.x timeline to align uncleaned timeline as active and rest of it as archived or a minimum of 2 days of timeline. 

For eg, 
Case 1: Someone configures cleaner to 4 days. 

1.x active timeline will be 4 days and rest will go into LSM 

Case2: Someone configures cleaner to 1 day:
1.x active timeline will be 2 days and rest will go into LSM.

If we maintain this semantics, with 0.16.x reader, we don't need to port LSM reader. just reading active timeline would suffice for all of snapshot, time travel and incremental queries.

### Timeline changes:
Let’s reiterate what we need to support w/ 0.16.0 reader.

### Timeline read of 1.x need to be supported
- Commit instants w/ completion time should be readable. HoodieInstant parsing logic to parse completion time should be ported over.
- We could ignore the completion time log file ordering semantics since we don’t plan to support it yet in 0.16.0. But reading should not break.
- Pending Clustering commits using “cluster” action should be readable in 0.16.0 reader.
- HoodieDefaultTimeline should be able to support both 0.x timeline and 1.x timeline. 
- Commit metadata in both avro and json should be supported. The challenging part here is, for every commit metadata, we might have to deserialize to avro and on exception try json. We could deduce the format using completion file name, but as per current code layering, deserialization methods does not know the file name( method takes byte[]). Similarly for clustering commits, unless we have some kind of watermark, we have to keep considering replace commits as well in the FSV building logic to ensure we do not miss any pending clustering commits.

Actual code changes: 

> - So for above mentioned reasons, we plan to introduce a new table config named “table version upgrade commit time” which will be referring to the latest commit which was written in 0.16.0 just before upgrading to 1.x. Once we have this commit time as water mark, which marks the upgrade in the timeline, even a 1.x reader should only account for older compatible reads up until the water mark. For eg, to deser pending replace commits for the interest of clustering can be dropped after the water mark. Once we get past that, only “cluster”ing pending commits can represent pending clustering writes. We could use the same watermark for commit metadata parsing as well. So, that at some point(after the water mark) we could safely assume entire timeline is only in avro (speaking from a 1.x FSV/reader standpoint) and not in json.

Author: update: we are thinking to avoid the using it for commit metadata deser. 
a 0.16.x reader will favor json and then fallback to avro. 
0.1.x reader will favor avro and then fallback json.

- b. Some of the methods were renamed in 1.x compared to 0.16.0.
findInstantsInRangeByStateTransitionTime -> findInstantsInRangeByCompletionTime
findInstantsModifiedAfterByStateTransitionTime -> findInstantsModifiedAfterByCompletionTime
getInstantsOrderedByStateTransitionTime -> getInstantsOrderedByCompletionTime
We need to add back these older methods to HoodieDefaultTimeline, so that we do not need to fix all the callers in 0.16.0. We could certainly introduce private methods and avoid code duplication.
- c. HoodieInstant changes from 1.x need to be ported over so that completion time is accounted for. If completion time is present, we should account for that, if not, we can fallback to last mod time of the completed commit meta file.
- d. We need to port code changes which accounts for uncommitted log files. In 0.16.0, from FSV standpoint, all log files(including partially failed) are valid. We let the log record reader ignore the partially failed log files. But
  in 1.x, log files could be rolledback (deleted) by a concurrent rollback. So, the FSV should ensure it ignores the uncommitted log files.

> - e. full LSM based read capability needs to be ported back to 0.16.x timeline reader. 

Author: we are revisiting this. may not be needed

- f. Given the requirement to port over above changes, we have two options to go about HoodieDefaultTimeline. But one option to consider: we could introduce Hoodie016xDefaultTimeline and Hoodie1xDefaultTimeline and use delegate pattern to delegate to either of the timelines. Using hoodie table version we could instantiate (internally to HoodieDefaultTimeline) to either of Hoodie016xDefaultTimeline or Hoodie1xDefaultTimeline. 
Or we need to make pointed fixes to entire HoodieDefaultTimeline to account for some of these backwards compatible logic. But given that 1.x reader should anyway support most of these backwards compatible ones, our proposal is to just have one HoodieDefaultTimeline(ported over from 1.x) and make pointed fixes to ensure its backwards compatible.  

### FileSystemView changes
Once all timeline changes are incorporated, we need to account for FSV changes. Major change as called out earlier is the Completion time based log files from 1.x writer and the log file naming referring to delta commit time instead of base commit time. So, w/o any changes to FSV/HoodieFileGroup/HoodieFileSlice code snippets, our file slice deduction logic might be wrong. 

> Each log file could be tagged as its own file slice since each has a different base commit time (thats how 0.16.x HoodieLogFile would deduce it). So, we might have to port over CompletionTimeQueryView class and associated logic to 0.16.0. So, for file slice deduction logic in 0.16.0 will be pretty much similar to 1.x reader. But the log file ordering for log reading purpose, we do not need to maintain parity with 1.x reader as of yet. (unless we make NBCC default with MDT).

Author: we don't need completion time query view. 

we will fetch the instant time from log file name (could be base instant time incase of 0.x, could be delta commit time in 1.x).
and we need to find the largest base instant time < this instant time

  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
# RFC-76: [Bridge release for 1.x]

## Proposers

- @nsivabalan
- @vbalaji

## Approvers
 - @yihua
 - @codope

## Status

JIRA: https://issues.apache.org/jira/browse/HUDI-7882

> Please keep the status updated in `rfc/README.md`.

## Abstract

[Hudi 1.x](https://github.com/apache/hudi/blob/ae1ee05ab8c2bd732e57bee11c8748926b05ec4b/rfc/rfc-69/rfc-69.md) is a powerful 
re-imagination of the transactional database layer in Hudi to power continued innovation across the community in the coming 
years. It introduces lot of differentiating features for Apache Hudi. Feel free to checkout the 
[release page](https://hudi.apache.org/releases/release-1.0.0-beta1) for more info. We had beta1 and beta2 releases which was meant for 
interested developers/users to give a spin on some of the  advanced features. But as we are working towards 1.0 GA, we are proposing 
a bridge release (0.16.0) for smoother migration for existing hudi users. 

## Objectives 

> Goal is to have a smooth migration experience for the users from 0.x to 1.0. We plan to have a 0.16.0 bridge release asking everyone to first migrate to 0.16.0 before they can upgrade to 1.x. 

Reviewer: ```suggestion
Goal is to have a smooth migration experience for the users from 0.x to 1.0. We plan to have a 0.16.0 bridge release asking users to first migrate writers (spark, flink, kafka connect sinks) and readers (Spark, Flink, Hive, Presto, Trino, ...) to 0.16.0 before they can upgrade to 1.x. 
```

Reviewer: lets be specific about readers/writers as much as possible.

> A typical organization might have a medallion architecture deployed to run 1000s of Hudi pipelines i.e. bronze, silver and gold layer. 

Reviewer: ```suggestion
A typical organization might have a medallion architecture deployed to run 1000s of Hudi pipelines i.e. bronze, silver and gold layer, where pipelines can be both readers and writers. 
```

> For this layout of pipelines, here is how a typical migration might look like(w/o a bridge release)

Reviewer: ```suggestion
For this layout of pipelines, here are some challenges faced during migration  (w/o a bridge release)
```

a. Existing pipelines are in 0.15.x. (bronze, silver, gold) 
b. Migrate gold pipelines to 1.x. 
- We need to strictly migrate only gold to 1x. Bcoz, a 0.15.0 reader may not be able to read 1.x hudi tables. So, if we migrate any of silver pipelines to 1.x before migrating entire gold layer, we might end up in a situation, 
where a 0.15.0 reader (gold) might end up reading 1.x table (silver). This might lead to failures. So, we have to follow certain order in which we migrate pipelines. 
c. Once all of gold is migrated to 1.x, we can move all of silver to 1.x. 

> d. Once all of gold and silver pipelines are migrated to 1.x, finally we can move all of bronze to 1.x.

Reviewer 3: like you mentioned, there might be some issues after upgrade, so the user wants to downgrade the 1.x table back to 0.x, and we got a scenario where the old reader needs to read the new files(or just force restore to where the old table commits ends)?

Author: hey @danny0405 : not sure what do you intend by this comment. 
here, I am mainly focussing a happy path migration just to give a glimpse of what a typical migration looks like. I guess current content L46 to 67 conveys that. I feel, we don't need to discuss the downgrade scenario here. let me know what do you think.

> In the end, we would have migrated all of existing hudi pipelines from 0.15.0 to 1.x. 

Reviewer: nit: tighten up language around key point - which is - we need a bridge release.

But as you could see, we need some coordination with which we need to migrate. And in a very large organization, sometimes we may not have good control over downstream consumers. 
Hence, coordinating entire migration workflow and orchestrating the same might be challenging.

Hence to ease the migration workflow for 1.x, we are introducing 0.16.0 as a bridge release.  

Here are the objectives with this bridge release:

> - 1.x reader should be able to read 0.14.x to 0.16.x tables w/o any loss in functionality and no data inconsistencies.

Reviewer: ```suggestion
- 1.x reader should be able to read 0.14.x to 0.16.x tables w/o any loss in 0.x functionality and no data inconsistencies.
```

> - 0.16.x should have read capability for 1.x tables w/ some limitations. For features ported over from 0.x, no loss in functionality should be guaranteed. 

Reviewer: worth calling out guarantees on "data inconsistencies" ?

But for new features that was introduced in 1.x, we may not be able to support all of them. Will be calling out which new features may not work with 0.16.x reader. 
- In this case, we explicitly request users to not turn on these features untill all readers are completely migrated to 1.x so as to not break any readers as applicable. 

> Connecting back to our example above, lets see how the migration might look like for an existing user. 

Reviewer: ```suggestion
Connecting back to our example above, lets see how the bridge release makes life easier for an existing user. 
```

a. Existing pipelines are in 0.15.x. (bronze, silver, gold)
b. Migrate pipelines to 0.16.0 (in any order. we do not have any constraints around which pipeline should be migrated first). 
c. Ensure all pipelines are in 0.16.0 (both readers and writers)
d. Start migrating pipelines in a rolling fashion to 1.x. At this juncture, we could have few pipelines in 1.x and few pipelines in 0.16.0. but since 0.16.x 
can read 1.x tables, we should be ok here. Just that do not enable new features like Non blocking concurrency control yet. 
e. Migrate all of 0.16.0 to 1.x version. 

> f. Once all readers and writers are in 1.x, we are good to enable any new features (like NBCC) with 1.x tables.

Reviewer: going back to danny's point above, can we also think through scenarios where 1.x writer needs to be downgraded to 0.16 due to some issue say? are these supported.. lets make these clearer, so users can plan ahead?

As you could see, company/org wide coordination to migrate gold before migrating silver or bronze is relaxed with the bridge release. Only requirement to keep a tab on, 
is to ensure to migrate all pipelines completely to 0.16.x before starting to migrate to 1.x.

> So, here are the objectives of this RFC with the bridge release. 

Reviewer: feels redundant? can we move/merge these bullets with the objectives stated above

Author: Above I have given justification for need for a bridge release.
here, I am listing what are the deliverables from this RFC. not sure if we can combine them. I have taken a stab

- 1.x reader should be able to read 0.14.x to 0.16.x tables w/o any loss in functionality and no data inconsistencies.
- 0.16.x should have read capability for 1.x tables w/ some limitations. For features ported over from 0.x, no loss in functionality should be guaranteed.
  But for new features that are being introduced in 1.x, we may not be able to support all of them. Will be calling out which new features may not work with 0.16.x reader.
- Document steps for rolling upgrade from 0.16.x to 1.x with minimal downtime.
- Downgrade from 1.x to 0.16.x documented with call outs on any functionality loss. 

> ### Considerations when choosing Migration strategy

Reviewer: nit: camel case

> - While migration is happening, we want to allow readers to continue reading data. This means, we cannot employ a stop-the-world strategy when we are migrating. 

Reviewer: lets document impact on incremental /cdc queries as well?

All the actions that we are performing as part of table upgrade should not have any side-effects of breaking snapshot isolation for readers.
- Also, users should have migrated to 0.16.x before upgrading to 1.x. We do not want to add read support for very old versions of hudi in 1.x(for eg 0.7.0). 
- So, in an effort to bring everyone to latest hudi versions, 1.x reader will have full read capabilities for 0.16.x, but for older hudi versions, 1.x reader may not have full reader support. 
The reocmmended guideline is to upgrade all readers and writers to 0.16.x. and then slowly start upgrading to 1.x(readers followed by writers). 

Before we dive in further, lets understand the format changes:

> ## Format changes

Reviewer: ```suggestion
## Format changes in 1.x
```

### Table properties
- Payload class ➝ payload type.
- hoodie.record.merge.mode is introduced in 1.x. 
- New metadata partitions could be added (optionally enabled)

> ### MDT changes

Reviewer: along with each change, some commentary on how this affects the migration would be helpful. 

For e.g new MDT partitions on 1.x writer, simply can remain unused by a 0.16 reader

- New MDT partitions are available in 1.x. MDT schema upgraded.
- RLI schema is upgraded to hold row positions.

### Timeline:
- [storage changes] Completed write commits have completed times in the file name(timeline commit files).
- [storage changes] Completed and inflight write commits are in avro format which were json in 0.x.
- We are switching the action type for pending clustering from “replace commit” to “cluster”.

> - [storage changes] Timeline ➝ LST timeline. There is no archived timeline in 1.x.

Reviewer: ```suggestion
- [storage changes] Timeline ➝ LSM timeline. There is an archived timeline in 1.x, but its been redesigned for scalable access
```

- [In-memory changes] HoodieInstant changes due to presence of completion time for completed HoodieInstants.

### Filegroup/FileSlice changes:
- Log file names contain delta commit time instead of base instant time.

> - Log appends are disabled in 1.x. In other words, each log block is already appended to a new log file.

Reviewer: ```suggestion
- Log appends are disabled in 1.x. In other words, each log block is already appended to a new log file. There is 1-1 mapping from a log block to log file
```

- File Slice determination logic for log files changed. In 0.x, we have base instant time in log files and its straight forward. 
In 1.x, we find completion time for a log file and find the base instant time (parsed from base files) for a given HoodieFileGroup, 
- which has the highest value lesser than the completion time of the log file) of interest. 
- Log file ordering within a file slice. (in 0.x, we use base instant time ➝ log file versions ➝ write token) to order diff log files. in 1.x, we will be using completion time to order).
- Rollbacks in 0.x appends a new rollback block (new log file). While in 1.x, rollback will remove the partially failed log files. 

### Log format changes:
- We have added a new header type, IS_PARTIAL in 1.x.

## Changes to be ported over to 0.16.x to support reading 1.x tables

### What will be supported

> - For features introduced in 0.x, and tables written in 1.x, 0.16.0 reader should be able to provide consistent reads w/o any breakage. 

Reviewer 3: So the under-mentioned 1.x features are excluded right?

Reviewer: lets expand this more?  for e,g cover types of tables, queries and writer/reader paths to be explicit.. This helps users read this RFC and figure out their plans.

> ### What will not be supported

Reviewer: I reviewed till here.. Resuming later tonight.

| Partial merges/updates  | supported | Fails with clear error message stating that partial merging |
| Functional indexes | supported | Not supported. Perf optimization may not kick in|
| Secondary indexes | supported | Not supported. Perf optimization may not kick in|
| NBCC OR Completion time based log file ordering in a file slice | supported | Not supported. Will be using log file version and write token based ordering.|  

### Timeline
- Commit instants w/ completion time should be readable. HoodieInstant parsing logic to parse completion time should be ported over.
- Commit metadata in avro instead of json should be readable.
   - More details on this under Implementation section.
- Pending Clustering commits using “cluster” action should be readable in 0.16.0 reader.
- HoodieDefaultTimeline should be able to support both 0.x timeline and 1.x timeline.
   - More details on this under Implementation section.
- Should we port LSM reader as well to 0.16.x? Our goal here is to support snapshot, time travel and incremental queries for 1.x tables. Strictly speaking we can only do all these 3 queries in uncleaned 

> instants i.e. over instant ranges where cleaner has not been executed. So, if we guarantee that 1.x active timeline will definitely contain all the uncleaned instants, we could get away by not even porting over 

Reviewer 3: I think most of the reader logic is based on the active timeline instants, and the 1.x archival strategy does not really change comparing to 0.x, but we may still need to read the LSM archived timeline if there are 1.x written log files in the table while the 0.x reader want to read them(for the completion time).

Author: we don't need completion time in 0.16.0 reader if we do not need to support NBCC. as per feedback, I am going to add a section in this RFC on bridge release design if we were to support NBCC as well. To get that working, we will be needing completion time query view.

LSM timeline reader logic to 0.16.0. 

#### Completion time based read in FileGroup/FileSlice grouping and log block reads
- As of this write up, we are not supporting completion time based log file ordering in in 0.16. Punting it as called out earlier.
- What's the impact if not for this support:
    - For users having single writer mode or OCC in 1.x, 0.16.0 reader not supporting completion time based read(log file ordering) should still be ok. Both 1.x reader and 0.16.0 reader should have same behavior.

>     - But only if someone has NBCC writes and have log files in different ordering written compared to the log file versions, 0.16.0 reader might result in data consistency issues. but since we are calling out that 0.16.0 is a bridge release and recommend users to migrate all the readers to 1.x fully before starting to enable any new features for 1.x tables.

Reviewer 3: sounds fair.

>     - Example scenarios. say, we have lf1_10_25, lf2_15_20(format "logfile[index]_[starttime]_[completiontime]") for a file slice. In 1.x reader, we will order and read it as lf2 followed by lf1. w/o this support in 0.16.0, we might read lf1 followed by lf2. Just to re-iterate this might only impact users who have enabled NBCC and having multi writers writing log files in different ordering. Even if they were using OCC, one of the writers is expected to have failed (on the writer side) since data is overlapping for two writers in 1.x writer.

Author: yes, you are right. or only when all readers are migrated to 1.x

### FileSystemView:

> - Support ignoring partially failed log files from FSV. In 0.16.0, from FSV standpoint, all log files(including partially failed) are valid. We let the log record reader ignore the partially failed log files. But

Author: yes. but wrt this line "We let the log record reader ignore the partially failed log files", I was referring to log record reader from 0.16.x.

> - Support ignoring partially failed log files from FSV. In 0.16.0, from FSV standpoint, all log files(including partially failed) are valid. We let the log record reader ignore the partially failed log files. But

Author: not sure I understand your point. can you please clarify.

>   in 1.x, log files could be rolledback (deleted) by a concurrent rollback. So, the FSV should ensure it ignores the uncommitted log files. 

Author: reason to quote "concurrent" is: This will only impact those queries where log file was present during query planning (i.e FSV building), but it was deleted by the time query was executed (due to rollback). 
If rollback was happening serially, or if every query lines up well, where the file was never deleted in b/w a planning and an execution, we should be good. 
hope you got the gist.

- We don't need the completion time logic ported over either for file slice determination nor for log file ordering.\
    - So, here is how File Slice determination will happen using 0.16.0 reader.\
      a. Read base files and assign to resp file groups. \

>       b. Read log files. Parse instant time from log file name (it could refer to base instant time for a file written in 0.16.x, or it could refer to delta commit in case of 1.x writer). Find largest (<=) base instant times in the corresponding file group

Author: yes. I did validate this for a test table.

      and assign the log file to it. \
      d. Log files within a file slice are ordered based on log version and write tokens.\

>       The same logic will be used whether we are reading a 0.16.x table or its a 1.x table. Only difference wrt how a 1.x reader will behave in comparison to 0.16.x reader while reading a 1.x table is when NBCC is involved w/ multi-writers. But as of this writing, 

Reviewer 3: In summary, there is no additional change for file slicing if we just port the 1.x file slicing into 0.16.x, it should work fine.

Author: relaying the same response I added above. 
```
we don't need completion time in 0.16.0 reader if we do not need to support NBCC. as per feedback, I am going to add a section in this RFC on bridge release design if we were to support NBCC as well. To get that working, we will be needing completion time query view.
```

  we are not supporting that in 0.16.x. 
    - Lets see with an illustrative example.\
      i. Table written in 0.16.x. Say total files are bf1_t10, lf1_t10, lf2_t10, bf2_t100, lf3_100, lf4_100.\
      If we run through above algo, lf1 and lf2 will be assigned to file slice with base instant time t10,\
  and lf3 and lf4 will be assigned to file slice with base instant time t100.\
      ii. Table written in 1.x.   Say total files are bf1_t10, lf1_t15_{t50}, lf2_t60_{t90}, bf2_t100, lf3_t110_{t140}, lf4_t150_{t200}. Time within braces are completion times which is not really part of log file name, but showing it just for illustration.\
      If we run through above algo, lf1 and lf2 will be assigned to file slice with base instant time t10 (since the max base instant time which is less than instant times of the log files (start time/delta commit time) is t10\
      and lf3 and lf4 will be assigned to file slice with base instant time t100 using similar logic.\

> - FSV building/reading should account for commit instants in both 0.x and 1.x formats. Since completion time log file ordering is not supported in 0.16.0 (that’s our current assumption), we may not need to support exactly how a FSV in 1.x reader supports. But below items should be supported.

Author: 1 minor correction to above logic. we need to fix log file ordering for a given file slice. for log files created in 1.x, since all of them have diff delta commits, the log file version is pretty much 1. So, our log file ordering should be based on "instant time" and then "log file version" and then "write token" in that order. 
This should take care of correct ordering for both 0.x log files and 1.x log files within a file slice.

Author: yes yes. we are not making any assumption. "instant time", "log file version", "write token" -> this is how our log file comparator will be designed

| FG1 having 3rd file slice in pending compaction and few log files added to latest file slice                                                          | latest file slice should be merged w/ log files from last but one file slice. log files are ordered based on completion times. | latest file slice should be merged w/ log files from last but one file slice. log files are ordered based on log version and write token |                                                                |
| Time travel Query                                                                                                                                     | works as expected                                                                                                              | Assuming commit time is in active timeline, we are good. Else, LSM timeline read support needs to be ported over.                        |

> | Incremental Query                                                                                                                                     | works as expected                                                                                                              | Assuming commit time is in active timeline, we are good. Else, LSM timeline read support needs to be ported over.                        |                                                                                       |

Reviewer 3: We are good for 0.16.x if we port the LSM timeline reader and the completion time query view.

Author: yes. to support NBCC, we need LSM timeline reader and completion time query view.

### Table properties
- payload type inference from payload class need to be ported.
- hoodie.record.merge.mode needs to be ported over. 

### MDT changes:
- MDT schema changes need to be ported.

### Log reader/format changes:
- New log header type need to be ported.
- For unsupported features, meaningful errors should be thrown. For eg, partial update has been enabled and if those log files are read using 0.16.0 reader, we should fail and throw a meaningful error.
- For deletion vector, we should add the fallback option to 0.16.0 so that we give up on perf optimization, but still support reads w/o any failures. 

### Read optimized query:
- No explicit changes required (apart from timeline, and FSV) to support read optimized query.  

### Incremental reads and time travel query: 

> - Incremental reads in 1.x is expected to have some changes and design is not fully out. So, until then we have to wait to design 0.16.x read support for 1.x tables for incremental queries. 

Reviewer: There are lots of users who are running incremental pipelines. We need to ensure the pipelines can be safely resume w.r.t checkpoints. reprocessing some commits again may be ok, but skipping is not (obv)

> - Time travel query: I don't see any changes in 1.x wrt time travel query. So, as per master, we are still using commit time(i.e not completed time) to serve time travel query. Until we make any changes to that, we do not need to add any additional support to 0.16.x reader. But if we plan to make changes in 1.x, we might have to revisit this. 

Reviewer: ```suggestion
- Time travel query: There are n't notable changes in 1.x wrt time travel query. We are still using commit time(i.e not completed time) to serve time travel query. Until we make any changes to that, we do not need to add any additional support to 0.16.x reader. But if we plan to make changes in 1.x, we might have to revisit this. 
```

Reviewer: we need to do this as well I think. So lets think based on this change happening.

any additional logic to 0.16.x reader just for the purpose of CDC in addition to changes covered in this RFC. 

## 0.16.0 ➝ 1.0 upgrade steps
This will be an automatic upgrade for users when they start using 1.x hudi library. Listing the changes we might need to do with the upgrade.
- Rewrite archival timeline to LSM based timeline.
- Do not touch active timeline, since there could be concurrent readers reading the table. So, 1.x reader should be capable of handling timeline w/ a mix of 0.x commit files and 1.x commit files.
- But we need to trigger rollback of any failed writes using 0.16.x rollback logic. Since with 1.x, rollback will delete the log files based on delta commit times in log file naming, this rollback logic may not work for 
log files written in 0.16.0, since we have log appends. So, we need to trigger rollbacks of failed writes using 0.16.x rollback flow and not 1.x flow. 
- No changes to log reader.
- Check custom payload class and hoodie.record.merge.mode in table properties and switch to respective 0.16.x properties.

> - Debatable: Trigger compaction for latest file slices. We do not want a single file slice having a mix of log files from 0.x and log files from 1.x. So, we will trigger a full compaction 

Reviewer: I suggest we do this.

of the table to ensure all latest file slices has just the base files. I need to dug deeper on this(need to have detailed discussion w/ initial authors who designed this). But if we ensure all partially failed writes are rollbacked completely, 
we should be good to upgrade w/o needing to trigger full compaction. 

Lets walk through an example:
Say we have a file group and latest file slice is fs3. and it has 2 log files in 0.x. 
  fs3(t3):
     lf1_(3_10) lf2_(3_10)
here format is lf[log version]_([committime]_[completiontime])
base file instant time is 3 and so all log files (in memory) has same begin and completion time.(Remember completion time is only applicable for commits and not for data files). and so, from 1.x reader standpoint,
lf1 and lf2 has delta commit time as t3 (which matches the base file's instant time) and its completion time is t10. 

lets trigger an upgrade and add 2 new files. 

fs3(t3):
lf1_(3_10) lf2_(3_10) [upgrade] lf3_15_20, lf4_40_100.

> With this layout, 1.x reader should be able to read this file slice correctly w/o any issues. So, we should not require a full compaction during upgrade. But need to consult w/ authors to ensure we are not missing anything.

Reviewer: I'd like to keep things simple. Its very likely that users are actively compacting  the table today anyway.

- We have two options wrt handling active timeline instants(for eg, pending clustering instants):
   A: No changes to active instants. In order to support 1.x tables w/ 0.16.0 reader, already these changes might have been ported to 0.16.0 reader. But we might have to ensure downgrade from 0.16.0 to any older version of hudi does take care of rewriting the active timeline instants.
   B: While downgrading from 1.x to 0.16.0, let’s rewrite the active timeline instants too. So that after downgrade, we can't differentiate whether table was natively created in 0.16.0 or was it upgraded to 1.x and then later downgraded. The con in Option A is not an issue here since we take care of active timeline instants during 1.x to 0.16.0 downgrade only. So, downgrade from 0.16.0 to any lower hudi version does not need to do any special handling if table was native to 0.16.0 or was it downgraded from 1.x to 0.16.0
   Our proposal is to go with Option B to ensure 0.16.0 table will be intact and not have any residues from 1.x writes. While the rewrite of active instants are happening, we could see some read failures. But 

>    we do not want to have a 0.16.0 (post upgrade) table to deal w/ 1.x timeline intricacies. 

Author: An optimization we thought that could help us in the downgrade flow or post downgrade flow: 
say we introduce a table property called "table.version.lineage", whose value is a comma separated list of all diff table versions this table has come through. 

for eg, "7,8". this means that, table was created w/ version 7 and then upgraded to 8. 

"7,8,7" means, table was with 7 and then upgraded to 8 and then downgraded to 7 again. 

so, by interception this config value we can deterministically say if table has been upgraded to 1.x before or not. If yes, we could completely avoid downgrading a 0.16.0 table further down (say 0.15.0.). 

If not, we should be allowing downgrade from 0.16.0 to 0.15.0. Reason is: say someone upgraded from 0.15.0 to 0.16.0 and due to some issues, they want to downgrade to 0.15.0. We should definitely allow them to downgrade. 

only sequence we want to add constraints is: after upgrading to 1.0.x, and downgraded to 0.16.0. After sometime, if they wish to downgrade to 0.15.0, we should avoid the downgrade.

Reviewer: I don't prefer introducing these properties into table props. This is something the operator has to be aware of.

Author: noted. but do you have good recommendation on how we can guard further downgrades then?

Case 1: 
table created with 0.15.0 and then upgraded to 0.16.0. 
user runs into some issue. just executes the downgrade script and it works. 

Case2: 
table created with 0.15.0 and then upgraded to 0.16.0 and then upgraded to 1.x 
Lot of commits succeeded creating file slices completely in 1.x. 
user runs into some issue. Downgrades from 1.x to 0.16.0 will trigger full compaction. But there could be older file slices which were completely written in 1.x which we are not going to do anything as of current RFC design. 
And later user downgrades to 0.15.0. Downgrade succeeds. 
and now when user tries time travel or incr query, it might break or give unexpected results since log files are written in 1.x. 

So, if not for above proposal, only other option I can think of is: we can intercept the presence of log files in 1.x while downgrading from 0.16.x to 0.15.x and throw exception claiming that we can't downgrade. i.e. atleast throw clear error message.

> - If there are new MDT partitions (functional index, sec index), nuke them. Update table properties.

Reviewer: ```suggestion
- If there are new MDT partitions (functional index, sec index), remove them. Update table properties.
```

> - Check latest snapshot files for log files w/ new features enabled. For eg, deletion vector, partial update. If any log files are found in latest snapshot, we have to trigger compaction for those file groups. This calls for a custom compaction strategy as well which compacts only for certain file groups.

Reviewer: can we just force compaction using 1.x writer again, convert timeline/MT and move on?

Author: sure.

>    - In order to reduce downtime to downgrade, we are inclining to do this only for snapshot reads. If there are older file slices and users are issuing time travel or incremental reads, and if they have any of the new features enabled (deletion vector, partial update), readers could break after downgrade.

Author: An optimization we thought that could help us in the downgrade flow or post downgrade flow: 
say we introduce a table property called "table.version.lineage", whose value is a comma separated list of all diff table versions this table has come through. 

for eg, "7,8". this means that, table was created w/ version 7 and then upgraded to 8. 

"7,8,7" means, table was with 7 and then upgraded to 8 and then downgraded to 7 again. 

so, by interception this config value we can deterministically say if table has been upgraded to 1.x before or not. If yes, we could completely avoid downgrading a 0.16.0 table further down (say 0.15.0.). 

If not, we should be allowing downgrade from 0.16.0 to 0.15.0. Reason is: say someone upgraded from 0.15.0 to 0.16.0 and due to some issues, they want to downgrade to 0.15.0. We should definitely allow them to downgrade. 

only sequence we want to add constraints is: after upgrading to 1.0.x, and downgraded to 0.16.0. After sometime, if they wish to downgrade to 0.15.0, we should avoid the downgrade.

## Implementation details

To ensure read capability of 1.x tables in 0.16.0, documenting the changes required at class level. Other two sections (0.16.0 upgrade to 1.x or 1.x downgrade to 0.16.0) should be fairly straightforward and so not calling out the obvious.

### Timeline changes:
Let’s reiterate what we need to support w/ 0.16.0 reader.

### Timeline read of 1.x need to be supported

> - Commit instants w/ completion time should be readable. HoodieInstant parsing logic to parse completion time should be ported over.

Reviewer: again : please avoid using java class styles in these docs. (just so that people with enough time on their hands, don't later claim this as proof of "tied to jvm" etc)

- We could ignore the completion time log file ordering semantics since we don’t plan to support it yet in 0.16.0. But reading should not break.

> - Pending Clustering commits using “cluster” action should be readable in 0.16.0 reader.

Author: NTR: 
As as optimization, we are planning to consider "table upgrade commit time" to help us here. because, just after upgrading to 1.x, we have to consider both pending replace commits and pending clustering commits (for clustering operations). But at some point eventually, we want to sunset checking replace commit timeline and only consider clustering timeline (1 month or 2 months or next release 1.0.1). But until then, this config "table upgrade commit time" will help us do deterministic processing. for eg, once active timeline's first entry is > "table upgrade commit time", its safe to consider only pending clustering timeline and ignore pending replace timeline.

- HoodieDefaultTimeline should be able to support both 0.x timeline and 1.x timeline. 

> - Commit metadata in both avro and json should be supported. The challenging part here is, for every commit metadata, we might have to deserialize to avro and on exception try json. We could deduce the format using completion file name, but as per current code layering, deserialization methods does not know the file name( method takes byte[]). Similarly for clustering commits, unless we have some kind of watermark, we have to keep considering replace commits as well in the FSV building logic to ensure we do not miss any pending clustering commits.

Author: NTR: 
one option we are considering here is: 
we can leverage the "table upgrade commit time". So, we could use that to determine whether we need to deser as json or avro deterministically.

> - Commit metadata in both avro and json should be supported. The challenging part here is, for every commit metadata, we might have to deserialize to avro and on exception try json. We could deduce the format using completion file name, but as per current code layering, deserialization methods does not know the file name( method takes byte[]). Similarly for clustering commits, unless we have some kind of watermark, we have to keep considering replace commits as well in the FSV building logic to ensure we do not miss any pending clustering commits.

Reviewer: My suggestion is to do simpler things, keeping migration logic at the cost of extra parsing etc (relatively negligible compute overhead)

Author: ok

Most of the timeline, FSV are shared b/w spark and flink. Only additional changes we did to Flink that is not in spark (yet) is the incremental read. We have introduced completion time based incremental query capabilities in 1.x Flink reads. 
So, we might have to port over the changes to 0.16.x reader. Depending on table version, we could go with either of the two implementations(0.x Incremental read, or 1.x Incremental reads). So, both implementations will reside in 0.16.0 depending on which table is being read. 
We do not want to make nit-picky changes and will keep two implementations separate.

#### Pending exploration:
1. We removed special suffixes to MDT operations in 1x. By skimming the code changes, we should be ok, i.e. no additional changes required. But we need to test the flow and flush out details if anything to be added to 0.16.x reader. 
2. Just during upgrade, if there are pending commits in timeline, if we trigger rollback, since rollback will follow 1.x flow, our rollback may not work (since we expect delta commit times in file names). But for log files written in 0.x,
we may not have delta commit times in 1.x So, we need to flush this out in finer detail. 

> ## Rollout/Adoption Plan

Reviewer: lets fill this out?

 - What impact (if any) will there be on existing users? 
 - If we are changing behavior how will we phase out the older behavior?
 - If we need special migration tools, describe them here.
 - When will we remove the existing behavior

## Test Plan

> Describe in few sentences how the RFC will be tested. How will we know that the implementation works as expected? How will we know nothing broke?.

Reviewer: ```suggestion
<Describe in few sentences how the RFC will be tested. How will we know that the implementation works as expected? How will we know nothing broke?.>
```

Reviewer: lets keep incomplete sections in a braces , so its clear its a WIP

- No changes to log reader.
- Check custom payload class in table properties and switch to payload type.
- Trigger compaction for latest file slices. We do not want a single file slice having a mix of log files from 0.x and log files from 1.x. So, we will trigger a full compaction 
of the table to ensure all latest file slices has just the base files. 
  - Lets dissect and see what it needs to support not requiring the full compaction. In general, we plan to add a table config to track the commit time (more on this later in this doc) when the upgrade was done. 

>     So, using the upgrade commit time, we should be able to use different log file comparator to order log files within a given file slice. 

Author: 0.16.x reader is not going to order log files based on completion time and will only be ordering based on log version even for 1.x tables. 
which means, even for a single file slice, having a mix of 0.x log files and 1.x log files, we should be good here. 

pending:
file slice determination:
HoodieLogFile.getBaseInstantTime() has to work for both log files (0.x log files and 1.x log files). if we ensure this is intact, we should be good.

On a high level, we need to ensure commit metadata in either format (avro or json) need to be supported. And “cluster” and completed “compaction”s need to be readable in 0.16.0 reader.
- But the challenging part is, for every commit metadata, we might have to deserialize to avro and on exception try json. We could deduce the format using completion file name, but as per current code layering, deserialization methods does not know the file name( method takes byte[]).
- Similarly for clustering commits, unless we have some kind of watermark, we have to keep considering replace commits as well in the FSV building logic to ensure we do not miss any clustering commits.
- To be decided: We also need to use diff LogFileComparators depending on the file slice's base instant time. If the file slices's base instant time is < table upgrade commit time, we use older log file comparator to order log files. but if file slice's base instant time > table upgrade commit time, we have to use new log file comparator (completion time). Tricky part is if a file slice contains a mix of log files. 

>  This fix definitely needs to go into 1.x, but whether we wanted to port this change to 0.16.x or not is yet to be discussed and decided. Lets zoom in a bit to see what will happen if a single file slice could contain a mix of log files using 1.x reader(this is a basic requirement to support 0.16.x tables in 1.x). 

Author: we need to fix 1.x reader to enforce completion time based log file ordering for file slice. after the fix, from our understanding, same logic should work for a file slice completely written in 0.x. bcoz, completion time will match for all log files. and then we should use log version to determine the ordering. 
we need to have lot of tests covering all these scenarios.

Author: which mean, the log file compartor logic from 1.x needs to be ported to 0.16.x reader

Assuming 1.x reader and 1.x FSV should be able to read data written in older hudi versions, we also have a potential option here for avoid making nit-picky changes similar to the option called out earlier.
We could instantiate two different FSV depending on the table version. If table version is 7 (0.16.0), we could instantiate FSV_V0 may be and if table version is 8 (1.0.0), we could instantiate FSV_V1. So that we don’t break/regress any of 0.16.0 read functionality in the interest of supporting 1.x table reads. We should strive to cover all scenarios and not let any bugs creep in, but trying to see if we can keep the changes isolated so that battle tested code (FSV) is not touched or changed for the purpose of supporting 1.x table reads. If we run into any bugs with 1.x reads, we could ask users to not upgrade any of the writers to 1.x and stick with 0.16.0 unless we have say 1.0.1 or something. But it would be really bad if we break 0.16.0 table read in some edge case.  Just calling out as one of the safe option to upgrade.

#### Pending exploration:

> 1. We removed special suffixes to MDT operations in 1x. we need to test the flow and flush out details if anything to be added to 0.16.x reader. 

Author: understand the new commit time generation logic is foolproof. what incase there is a concurrent ingestion in data table co-incidentally generates the same commit time?
